Ray is incompatible with pydantic>=2.0

We should pin `pydantic<2.0` to workaround the issues on Ray side.


Example [from](https://github.com/modin-project/modin/actions/runs/5425526005/jobs/9866377841):
```bash
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/ray/util/state/__init__.py", line 1, in <module>
    from ray.util.state.api import (
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/ray/util/state/api.py", line 17, in <module>
    from ray.util.state.common import (
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/ray/util/state/common.py", line 120, in <module>
    @dataclass(init=True)
  File "/usr/share/miniconda/envs/modin/lib/python3.8/site-packages/pydantic/dataclasses.py", line 139, in dataclass
    assert init is False, 'pydantic.dataclasses.dataclass only supports init=False'
AssertionError: pydantic.dataclasses.dataclass only supports init=False
```

--- a/docs/requirements-doc.txt
+++ b/docs/requirements-doc.txt
@@ -14,6 +14,8 @@ sphinx<6.0.0
 sphinx-click
 # ray==2.5.0 broken: https://github.com/conda-forge/ray-packages-feedstock/issues/100
 ray[default]>=1.13.0,!=2.5.0
+# https://github.com/modin-project/modin/issues/6336
+pydantic<2
 # Override to latest version of modin-spreadsheet
 git+https://github.com/modin-project/modin-spreadsheet.git@49ffd89f683f54c311867d602c55443fb11bf2a5
 sphinxcontrib_plantuml

--- a/environment-dev.yml
+++ b/environment-dev.yml
@@ -6,6 +6,8 @@ dependencies:
   - numpy>=1.18.5
   # ray==2.5.0 broken: https://github.com/conda-forge/ray-packages-feedstock/issues/100
   - ray-default>=1.13.0,!=2.5.0
+  # https://github.com/modin-project/modin/issues/6336
+  - pydantic<2
   - pyarrow
   # workaround for https://github.com/conda/conda/issues/11744
   - grpcio!=1.45.*

--- a/examples/tutorial/jupyter/execution/pandas_on_ray/requirements.txt
+++ b/examples/tutorial/jupyter/execution/pandas_on_ray/requirements.txt
@@ -3,4 +3,6 @@ jupyterlab
 ipywidgets
 tqdm
 modin[ray]
+# https://github.com/modin-project/modin/issues/6336
+pydantic<2
 modin[spreadsheet]

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -4,6 +4,8 @@ dask[complete]>=2.22.0
 distributed>=2.22.0
 # ray==2.5.0 broken: https://github.com/conda-forge/ray-packages-feedstock/issues/100
 ray[default]>=1.13.0,!=2.5.0
+# https://github.com/modin-project/modin/issues/6336
+pydantic<2
 pyarrow
 psutil
 fsspec

--- a/setup.py
+++ b/setup.py
@@ -6,7 +6,8 @@ with open("README.md", "r", encoding="utf-8") as fh:
 
 dask_deps = ["dask>=2.22.0", "distributed>=2.22.0"]
 # ray==2.5.0 broken: https://github.com/conda-forge/ray-packages-feedstock/issues/100
-ray_deps = ["ray[default]>=1.13.0,!=2.5.0", "pyarrow"]
+# pydantic<2: https://github.com/modin-project/modin/issues/6336
+ray_deps = ["ray[default]>=1.13.0,!=2.5.0", "pyarrow", "pydantic<2"]
 unidist_deps = ["unidist[mpi]>=0.2.1"]
 remote_deps = ["rpyc==4.1.5", "cloudpickle", "boto3"]
 spreadsheet_deps = ["modin-spreadsheet>=0.1.0"]

